### ISSUE TITLE
[Bug] Multiple batches fails in `quantized` example

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/candle#2072]
Hello,

Thank you for your amazing work here. We noticed this crash when testing parallel performance in Mistral.rs I was able to create a minimum reproducible example by running the quantized example and modified it to run 2 batches at once:
`cargo run --release --features cuda --example quantized -- --prompt "Why is the sky blue"`

```diff
for index in 0..to_sample {
    let input = Tensor::new(&[next_token], &device)?.unsqueeze(0)?;
+   let input = Tensor::cat(&[&input, &input], 0)?;
    let logits = model.forward(&input, prompt_tokens.len() + index)?;
    let logits = logits.squeeze(0)?;
    ...
```

However, I got an error:
"Error: unexpected y size 8192, 4096 1"

I think it comes from this line:
> https://github.com/huggingface/candle/blob/af955f260cc20364b3e000c895dcb134a46e4e94/candle-core/src/quantized/cuda.rs#L177-L179

I would appreciate any help!

**Comment by LaurentMazare** [comment]
Ah thanks for reporting this, it's not the line that you mention but there is definitely a bug with the batch dim in the newly added quantized matmul, I've put a fix in #2073 but will also add support for this in mat-mul-vec so it will be a bit longer (+ I'll have to add some tests).

**Comment by EricLBuehler** [comment]
@LaurentMazare, thank you for getting a fix so quickly! However, I still have an issue:

`Error: shape mismatch in cat for dim 0, shape for arg 1: [1, 32, 6, 128] shape for arg 2: [2, 32, 1, 128]`

**Comment by EricLBuehler** [comment]
@LaurentMazare, thank you for fixing this so quickly!
